OpenRA.Mods.Common/Traits/BotModules/PowerDownBotManager.cs

```diff
@@ -0,0 +1,201 @@
+		readonly Func<Actor, bool> isTogglableValid;
+
+		PowerManager playerPower;
+		int toggleTick;
+
+		// We keep a list to track toggled buildings for performance.
+		List<ToggledPowerWrapper> toggled = [];
+
+		sealed class ToggledPowerWrapper
+		{
+			public int ExpectedPowerChanging;
+			public Actor Actor;
+
+			public ToggledPowerWrapper(Actor a, int p)
+			{
+				Actor = a;
+				ExpectedPowerChanging = p;
+			}
+		}
+
+		public PowerDownBotModule(Actor self, PowerDownBotModuleInfo info)
+			: base(info)
+		{
+			world = self.World;
+			player = self.Owner;
+			togglable = new ActorIndex.OwnerAndNamesAndTrait<ToggleConditionOnOrderInfo>(world, info.PowerDownTypes, player);
+
+			isTogglableValid = a => a != null && a.Owner == self.Owner && !a.IsDead && a.IsInWorld;
+		}
+
+		protected override void Created(Actor self)
+		{
+			playerPower = self.Owner.PlayerActor.TraitOrDefault<PowerManager>();
+		}
+
+		protected override void TraitEnabled(Actor self)
+		{
+			toggleTick = world.LocalRandom.Next(Info.Interval);
+		}
+
+		// We calculate the approximate power changing if toggled on. Since we have no idea on the conditions set by users
+		// Calculation here need user to use PowerMultiplier on toggle control only on related buildings, for calculation of this bot module.
+		static int GetTogglePowerChanging(Actor a)
+		{
+			var powerChangingIfToggled = 0;
+			var powerTraits = a.TraitsImplementing<Power>().Where(t => !t.IsTraitDisabled).ToArray();
+			if (powerTraits.Length > 0)
+			{
+				var powerMulTraits = a.TraitsImplementing<PowerMultiplier>().ToArray();
+				powerChangingIfToggled = powerTraits.Sum(p => p.Info.Amount) * (powerMulTraits.Sum(p => p.Info.Modifier) - 100) / 100;
+				if (Array.Exists(powerMulTraits, t => !t.IsTraitDisabled))
+					powerChangingIfToggled = -powerChangingIfToggled;
+			}
+
+			return powerChangingIfToggled;
+		}
+
+		IEnumerable<ToggledPowerWrapper> GetOnlineBuildings()
+		{
+			var toggleableBuildings = new List<ToggledPowerWrapper>();
+
+			foreach (var a in togglable.Actors.Where(a => !a.IsDead && a.Info.HasTraitInfo<PowerInfo>()))
+			{
+				// Note: it is OK if GetTogglePowerChanging is not accurate, when player is still in lowpower.
+				// The bot will try to toggle off more buildings next bot tick.
+				var powerChanging = GetTogglePowerChanging(a);
+				if (powerChanging > 0)
+					toggleableBuildings.Add(new ToggledPowerWrapper(a, powerChanging));
+			}
+
+			return toggleableBuildings.OrderBy(bpw => bpw.ExpectedPowerChanging);
+		}
+
+		void IBotTick.BotTick(IBot bot)
+		{
+			if (toggleTick > 0 || playerPower == null)
+			{
+				toggleTick--;
+				return;
+			}
+
+			var power = playerPower.ExcessPower;
+			var togglingBuildings = new List<Actor>();
+
+			// When there is extra power, check if AI can toggle on
+			if (power > 0)
+			{
+				toggled = toggled.Where(bpw => isTogglableValid(bpw.Actor)).OrderByDescending(bpw => bpw.ExpectedPowerChanging).ToList();
```

Comment:
```suggestion
			if (power > 0 && toggled.Count > 0)
			{
				toggled.RemoveAll(bpw => !isTogglableValid(bpw.Actor));
				toggled.Sort((bpw1, bpw2) => bpw2.ExpectedPowerChanging.CompareTo(bpw1.ExpectedPowerChanging));
```

Maybe let's reuse the collection?

Reply:
kk, then it can be readobly